vaadin-combo-box-flow/src/test/java/com/vaadin/flow/component/combobox/dataview/ComboBoxDataViewTest.java

```diff
@@ -0,0 +1,309 @@
+        // eagerly with in-memory data provider
+        comboBox.setOpened(true);
+
+        // We expect that the current implementation of 'setItems' with IMDP
+        // will delegate to 'setItems(DataProvider)'
+        Mockito.verify(comboBox).setItems(dataProviderCaptor.capture());
+
+        // Verify the predicate filter always returns true and passes all items
+        Assert.assertArrayEquals(items, dataView.getItems().toArray());
+
+        // Now set the predicate and verify it goes to query parameter
+        inMemoryDataProvider.setFilter(item -> item.length() == 6);
+        Assert.assertArrayEquals(new String[] { "banana", "iguana" },
+                dataView.getItems().toArray());
+
+        // Finally set the text filter to the ComboBox and fetch the items
+        DataProvider<String, String> genericDataProvider = dataProviderCaptor
+                .getValue();
+
+        List<String> filteredItems = genericDataProvider
+                .fetch(new Query<>(0, Integer.MAX_VALUE, null, null, "ba"))
+                .collect(Collectors.toList());
+
+        // The result should contain an intersection of two filters (text and
+        // predicate), so the result should start with 'ba' and contain 6 chars.
+        Assert.assertEquals(1, filteredItems.size());
+        Assert.assertEquals("banana", filteredItems.get(0));
+    }
+
+    @Test
+    public void setInMemoryDataProviderWithNoConverter_throws() {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException
+                .expectMessage(String.format("ComboBox does not support "
+                        + "setting a custom in-memory data provider without "
+                        + "knowledge of the rules on how to convert internal text filter "
+                        + "into a predicate applied to the data provider. Please use%n"
+                        + "setItems(InMemoryDataProvider<T>, SerializableFunction<String, "
+                        + "SerializablePredicate<T>>)"
+                        + "%noverloaded method instead"));
+        component.setItems(dataProvider);
+    }
+
+    @Test
+    public void addItemCountChangeListener_serverSideFilterAlwaysEnabled() {
+        Assert.assertTrue(Boolean.parseBoolean(getClientSideFilter()));
+
+        dataView.addItemCountChangeListener(event -> {});
+
+        Assert.assertTrue(Boolean.parseBoolean(getClientSideFilter()));
+    }
+
+    @Test
+    public void addItemCountChangeListener_serverSideFilterEnabled() {
+        Assert.assertTrue(Boolean.parseBoolean(getClientSideFilter()));
+
+        dataView.addItemCountChangeListener(event -> {});
+
+        // Trigger ComboBox::refreshAllData()
+        dataProvider.refreshAll();
+
+        Assert.assertFalse(Boolean.parseBoolean(getClientSideFilter()));
+    }
+
+    @Test
+    public void addItemCountChangeListener_removeListeners_serverSideFilterDisabled() {
```

Comment:
ComboBox has a feature to not call the server side for fetching the items if the items count < page size:

```
setClientSideFilter(!forceServerSideFiltering && getDataProvider()
                .size(new Query<>()) <= getPageSizeDouble());
```

In that case the filtering is done on client side only, that means `flush()` method is never invoked = listeners are never notified. But if the developer adds an item count listener, then it's good to notify these listeners anyway, even if < page Size.

Reply:
This code has been removed. Another issue is that the client filter change should not trigger item count change event on the server. But this will be a separate ticket.